# Lab 2

In [2]:
import warnings
warnings.simplefilter('ignore')

import pandas as pd

df = pd.read_csv('s3://ia340-fall-2025-04052005/Crime_Data_from_2020_to_Present.csv') # load the data from s3
df.head() #show the top 5 rows or .tail() to show bottom 5

DR_NO               Date Rptd                DATE OCC  TIME OCC  AREA  \
0  211507896  04/11/2021 12:00:00 AM  11/07/2020 12:00:00 AM       845    15   
1  201516622  10/21/2020 12:00:00 AM  10/18/2020 12:00:00 AM      1845    15   
2  240913563  12/10/2024 12:00:00 AM  10/30/2020 12:00:00 AM      1240     9   
3  210704711  12/24/2020 12:00:00 AM  12/24/2020 12:00:00 AM      1310     7   
4  201418201  10/03/2020 12:00:00 AM  09/29/2020 12:00:00 AM      1830    14   

     AREA NAME  Rpt Dist No  Part 1-2  Crm Cd  \
0  N Hollywood         1502         2     354   
1  N Hollywood         1521         1     230   
2     Van Nuys          933         2     354   
3     Wilshire          782         1     331   
4      Pacific         1454         1     420   

                                         Crm Cd Desc  ... Status  Status Desc  \
0                                  THEFT OF IDENTITY  ...     IC  Invest Cont   
1     ASSAULT WITH DEADLY WEAPON, AGGRAVATED ASSAULT  ...     IC  Invest Cont   
2                                  THEFT OF IDENTITY  ...     IC  Invest Cont   
3  THEFT FROM MOTOR VEHICLE - GRAND ($950.01 AND ...  ...     IC  Invest Cont   
4    THEFT FROM MOTOR VEHICLE - PETTY ($950 & UNDER)  ...     IC  Invest Cont   

  Crm Cd 1 Crm Cd 2  Crm Cd 3 Crm Cd 4  \
0    354.0      NaN       NaN      NaN   
1    230.0      NaN       NaN      NaN   
2    354.0      NaN       NaN      NaN   
3    331.0      NaN       NaN      NaN   
4    420.0      NaN       NaN      NaN   

                                   LOCATION Cross Street      LAT       LON  
0   7800    BEEMAN                       AV          NaN  34.2124 -118.4092  
1           ATOLL                        AV     N  GAULT  34.1993 -118.4203  
2  14600    SYLVAN                       ST          NaN  34.1847 -118.4509  
3   6000    COMEY                        AV          NaN  34.0339 -118.3747  
4                   4700    LA VILLA MARINA          NaN  33.9813 -118.4350  

[5 rows x 28 columns]

In [3]:
df.shape #show the number of rows and columns 

(1004991, 28)

In [10]:
df.info() # shows the data types and number of null values. 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1004991 entries, 0 to 1004990
Data columns (total 28 columns):
 #   Column          Non-Null Count    Dtype  
---  ------          --------------    -----  
 0   DR_NO           1004991 non-null  int64  
 1   Date Rptd       1004991 non-null  object 
 2   DATE OCC        1004991 non-null  object 
 3   TIME OCC        1004991 non-null  int64  
 4   AREA            1004991 non-null  int64  
 5   AREA NAME       1004991 non-null  object 
 6   Rpt Dist No     1004991 non-null  int64  
 7   Part 1-2        1004991 non-null  int64  
 8   Crm Cd          1004991 non-null  int64  
 9   Crm Cd Desc     1004991 non-null  object 
 10  Mocodes         853372 non-null   object 
 11  Vict Age        1004991 non-null  int64  
 12  Vict Sex        860347 non-null   object 
 13  Vict Descent    860335 non-null   object 
 14  Premis Cd       1004975 non-null  float64
 15  Premis Desc     1004403 non-null  object 
 16  Weapon Used Cd  327247 non-null   fl

In [13]:
df.describe() #provides descriptive statistics

DR_NO      TIME OCC          AREA   Rpt Dist No      Part 1-2  \
count  1.004991e+06  1.004991e+06  1.004991e+06  1.004991e+06  1.004991e+06   
mean   2.202215e+08  1.339900e+03  1.069174e+01  1.115633e+03  1.400348e+00   
std    1.319718e+07  6.510613e+02  6.110255e+00  6.111605e+02  4.899691e-01   
min    8.170000e+02  1.000000e+00  1.000000e+00  1.010000e+02  1.000000e+00   
25%    2.106169e+08  9.000000e+02  5.000000e+00  5.870000e+02  1.000000e+00   
50%    2.209159e+08  1.420000e+03  1.100000e+01  1.139000e+03  1.000000e+00   
75%    2.311103e+08  1.900000e+03  1.600000e+01  1.613000e+03  2.000000e+00   
max    2.521041e+08  2.359000e+03  2.100000e+01  2.199000e+03  2.000000e+00   

             Crm Cd      Vict Age     Premis Cd  Weapon Used Cd      Crm Cd 1  \
count  1.004991e+06  1.004991e+06  1.004975e+06   327247.000000  1.004980e+06   
mean   5.001568e+02  2.891706e+01  3.056201e+02      363.955300  4.999174e+02   
std    2.052731e+02  2.199272e+01  2.193021e+02      123.734528  2.050736e+02   
min    1.100000e+02 -4.000000e+00  1.010000e+02      101.000000  1.100000e+02   
25%    3.310000e+02  0.000000e+00  1.010000e+02      311.000000  3.310000e+02   
50%    4.420000e+02  3.000000e+01  2.030000e+02      400.000000  4.420000e+02   
75%    6.260000e+02  4.400000e+01  5.010000e+02      400.000000  6.260000e+02   
max    9.560000e+02  1.200000e+02  9.760000e+02      516.000000  9.560000e+02   

           Crm Cd 2     Crm Cd 3   Crm Cd 4           LAT           LON  
count  69160.000000  2314.000000   64.00000  1.004991e+06  1.004991e+06  
mean     958.101258   984.015990  991.21875  3.399821e+01 -1.180909e+02  
std      110.354348    52.350982   27.06985  1.610713e+00  5.582386e+00  
min      210.000000   310.000000  821.00000  0.000000e+00 -1.186676e+02  
25%      998.000000   998.000000  998.00000  3.401470e+01 -1.184305e+02  
50%      998.000000   998.000000  998.00000  3.405890e+01 -1.183225e+02  
75%      998.000000   998.000000  998.00000  3.416490e+01 -1.182739e+02  
max      999.000000   999.000000  999.00000  3.433430e+01  0.000000e+00

### In my data set, there are 1004991 rows and 28 columns seen. Columns like Date Occured, Area, Time Occured, Type of crime exist. This data set seems to be very well put together and contians a lot of useful information.

In [15]:
#rename columns
df = df.rename(columns={"Crm Cd Desc": "Type Crm"}) 
df.head()

DR_NO               Date Rptd                DATE OCC  TIME OCC  AREA  \
0  211507896  04/11/2021 12:00:00 AM  11/07/2020 12:00:00 AM       845    15   
1  201516622  10/21/2020 12:00:00 AM  10/18/2020 12:00:00 AM      1845    15   
2  240913563  12/10/2024 12:00:00 AM  10/30/2020 12:00:00 AM      1240     9   
3  210704711  12/24/2020 12:00:00 AM  12/24/2020 12:00:00 AM      1310     7   
4  201418201  10/03/2020 12:00:00 AM  09/29/2020 12:00:00 AM      1830    14   

     AREA NAME  Rpt Dist No  Part 1-2  Crm Cd  \
0  N Hollywood         1502         2     354   
1  N Hollywood         1521         1     230   
2     Van Nuys          933         2     354   
3     Wilshire          782         1     331   
4      Pacific         1454         1     420   

                                            Type Crm  ... Status  Status Desc  \
0                                  THEFT OF IDENTITY  ...     IC  Invest Cont   
1     ASSAULT WITH DEADLY WEAPON, AGGRAVATED ASSAULT  ...     IC  Invest Cont   
2                                  THEFT OF IDENTITY  ...     IC  Invest Cont   
3  THEFT FROM MOTOR VEHICLE - GRAND ($950.01 AND ...  ...     IC  Invest Cont   
4    THEFT FROM MOTOR VEHICLE - PETTY ($950 & UNDER)  ...     IC  Invest Cont   

  Crm Cd 1 Crm Cd 2  Crm Cd 3 Crm Cd 4  \
0    354.0      NaN       NaN      NaN   
1    230.0      NaN       NaN      NaN   
2    354.0      NaN       NaN      NaN   
3    331.0      NaN       NaN      NaN   
4    420.0      NaN       NaN      NaN   

                                   LOCATION Cross Street      LAT       LON  
0   7800    BEEMAN                       AV          NaN  34.2124 -118.4092  
1           ATOLL                        AV     N  GAULT  34.1993 -118.4203  
2  14600    SYLVAN                       ST          NaN  34.1847 -118.4509  
3   6000    COMEY                        AV          NaN  34.0339 -118.3747  
4                   4700    LA VILLA MARINA          NaN  33.9813 -118.4350  

[5 rows x 28 columns]

In [16]:
df.isnull().sum() #show the number of rows containing null value

DR_NO                   0
Date Rptd               0
DATE OCC                0
TIME OCC                0
AREA                    0
AREA NAME               0
Rpt Dist No             0
Part 1-2                0
Crm Cd                  0
Type Crm                0
Mocodes            151619
Vict Age                0
Vict Sex           144644
Vict Descent       144656
Premis Cd              16
Premis Desc           588
Weapon Used Cd     677744
Weapon Desc        677744
Status                  1
Status Desc             0
Crm Cd 1               11
Crm Cd 2           935831
Crm Cd 3          1002677
Crm Cd 4          1004927
LOCATION                0
Cross Street       850755
LAT                     0
LON                     0
dtype: int64

In [17]:
df = df.dropna() #drop null values
df.isnull().sum()

DR_NO             0
Date Rptd         0
DATE OCC          0
TIME OCC          0
AREA              0
AREA NAME         0
Rpt Dist No       0
Part 1-2          0
Crm Cd            0
Type Crm          0
Mocodes           0
Vict Age          0
Vict Sex          0
Vict Descent      0
Premis Cd         0
Premis Desc       0
Weapon Used Cd    0
Weapon Desc       0
Status            0
Status Desc       0
Crm Cd 1          0
Crm Cd 2          0
Crm Cd 3          0
Crm Cd 4          0
LOCATION          0
Cross Street      0
LAT               0
LON               0
dtype: int64

### The Column Crm Cd Desc was changed to Type Crm. Several Null values from the victims sex as well as the weapon used were identified then removed.

In [6]:
df['AREA'] #select a column 

0          15
1          15
2           9
3           7
4          14
           ..
1004986    21
1004987     4
1004988    13
1004989    17
1004990    19
Name: AREA, Length: 1004991, dtype: int64

In [18]:
df[5:10] #slice by rows

DR_NO               Date Rptd                DATE OCC  TIME OCC  \
537834  221718232  12/25/2022 12:00:00 AM  12/25/2022 12:00:00 AM      1150   
586003  221401314  11/10/2022 12:00:00 AM  11/10/2022 12:00:00 AM      2117   
728576  231717599  11/15/2023 12:00:00 AM  11/15/2023 12:00:00 AM       400   
809419  231915572  10/21/2023 12:00:00 AM  10/21/2023 12:00:00 AM         1   
934749  241905348  02/04/2024 12:00:00 AM  02/03/2024 12:00:00 AM      1100   

        AREA   AREA NAME  Rpt Dist No  Part 1-2  Crm Cd         Type Crm  ...  \
537834    17  Devonshire         1797         1     122  RAPE, ATTEMPTED  ...   
586003    14     Pacific         1452         2     910       KIDNAPPING  ...   
728576    17  Devonshire         1738         1     210          ROBBERY  ...   
809419    19     Mission         1902         1     210          ROBBERY  ...   
934749    19     Mission         1983         1     820  ORAL COPULATION  ...   

       Status   Status Desc Crm Cd 1 Crm Cd 2  Crm Cd 3 Crm Cd 4  \
537834     AA  Adult Arrest    122.0    230.0     910.0    998.0   
586003     IC   Invest Cont    812.0    860.0     910.0    998.0   
728576     IC   Invest Cont    210.0    230.0     761.0    998.0   
809419     AA  Adult Arrest    210.0    250.0     761.0    998.0   
934749     AO   Adult Other    761.0    820.0     910.0    998.0   

                               LOCATION                     Cross Street  \
537834  PARTHENIA                    ST                      HAYVENHURST   
586003                       WASHINGTON                         SPEEDWAY   
728576  HASKELL                      AV  SAN FERNANDO                 BL   
809419                             POLK                           BORDEN   
934749                           BURNET                        PARTHENIA   

            LAT       LON  
537834  34.2285 -118.4939  
586003  33.9792 -118.4666  
728576  34.2692 -118.4789  
809419  34.3103 -118.4467  
934749  34.2282 -118.4633  

[5 rows x 28 columns]

In [7]:
df.sort_values(by='Crm Cd Desc') #sort data by a column

DR_NO               Date Rptd                DATE OCC  TIME OCC  \
709797  230515242  10/30/2023 12:00:00 AM  10/30/2023 12:00:00 AM       800   
282575  210108514  04/07/2021 12:00:00 AM  04/07/2021 12:00:00 AM       839   
568531  221407662  03/16/2022 12:00:00 AM  03/15/2022 12:00:00 AM      2100   
770494  231507203  03/18/2023 12:00:00 AM  03/17/2023 12:00:00 AM       255   
483207  220315512  08/22/2022 12:00:00 AM  08/20/2022 12:00:00 AM      2315   
...           ...                     ...                     ...       ...   
622027  220717413  12/05/2022 12:00:00 AM  12/05/2022 12:00:00 AM      1210   
164613  201216692  07/18/2020 12:00:00 AM  07/18/2020 12:00:00 AM        25   
84259   202006611  01/17/2020 12:00:00 AM  01/16/2020 12:00:00 AM      2000   
110727  200808185  04/10/2020 12:00:00 AM  04/06/2020 12:00:00 AM      2030   
666840  240604490  01/11/2024 12:00:00 AM  11/30/2023 12:00:00 AM       700   

        AREA    AREA NAME  Rpt Dist No  Part 1-2  Crm Cd  \
709797     5       Harbor          566         1     648   
282575     1      Central          111         1     648   
568531    14      Pacific         1412         1     648   
770494    15  N Hollywood         1526         1     648   
483207     3    Southwest          334         1     648   
...      ...          ...          ...       ...     ...   
622027     7     Wilshire          721         2     756   
164613    12  77th Street         1248         2     756   
84259     20      Olympic         2088         2     756   
110727     8      West LA          833         2     756   
666840     6    Hollywood          668         2     756   

                       Crm Cd Desc  ... Status   Status Desc Crm Cd 1  \
709797                       ARSON  ...     IC   Invest Cont    648.0   
282575                       ARSON  ...     IC   Invest Cont    648.0   
568531                       ARSON  ...     IC   Invest Cont    648.0   
770494                       ARSON  ...     IC   Invest Cont    648.0   
483207                       ARSON  ...     AA  Adult Arrest    648.0   
...                            ...  ...    ...           ...      ...   
622027  WEAPONS POSSESSION/BOMBING  ...     IC   Invest Cont    756.0   
164613  WEAPONS POSSESSION/BOMBING  ...     IC   Invest Cont    756.0   
84259   WEAPONS POSSESSION/BOMBING  ...     IC   Invest Cont    756.0   
110727  WEAPONS POSSESSION/BOMBING  ...     IC   Invest Cont    756.0   
666840  WEAPONS POSSESSION/BOMBING  ...     AA  Adult Arrest    756.0   

       Crm Cd 2  Crm Cd 3 Crm Cd 4                                 LOCATION  \
709797      NaN       NaN      NaN  1100 S  PALOS VERDES                 ST   
282575      NaN       NaN      NaN   600 N  SPRING                       ST   
568531      NaN       NaN      NaN          WESTMINSTER                  AV   
770494      NaN       NaN      NaN  6200    BECK                         AV   
483207      NaN       NaN      NaN  3000    CRENSHAW                     BL   
...         ...       ...      ...                                      ...   
622027    998.0       NaN      NaN  8700    GRACIE ALLEN                 DR   
164613    998.0       NaN      NaN   900 W  64TH                         ST   
84259       NaN       NaN      NaN       W  WASHINGTON                   BL   
110727      NaN       NaN      NaN  1300    MIDVALE                      AV   
666840      NaN       NaN      NaN                               ST ANDREWS   

                           Cross Street      LAT       LON  
709797                              NaN  33.7342 -118.2814  
282575                              NaN  34.0582 -118.2387  
568531                             MAIN  33.9901 -118.4726  
770494                              NaN  34.1830 -118.3834  
483207                              NaN  34.0276 -118.3351  
...                                 ...      ...       ...  
622027                              NaN  34.0753 -118.3795  
164613                      

In [19]:
df[df['TIME OCC']>1000] # select rows by values 

DR_NO               Date Rptd                DATE OCC  TIME OCC  \
27751   201904032  01/02/2020 12:00:00 AM  01/01/2020 12:00:00 AM      2135   
134755  200613424  08/02/2020 12:00:00 AM  08/02/2020 12:00:00 AM      2030   
363765  210617136  10/08/2021 12:00:00 AM  10/07/2021 12:00:00 AM      1950   
490101  220600626  04/27/2022 12:00:00 AM  04/23/2022 12:00:00 AM      2300   
537834  221718232  12/25/2022 12:00:00 AM  12/25/2022 12:00:00 AM      1150   
586003  221401314  11/10/2022 12:00:00 AM  11/10/2022 12:00:00 AM      2117   
934749  241905348  02/04/2024 12:00:00 AM  02/03/2024 12:00:00 AM      1100   

        AREA   AREA NAME  Rpt Dist No  Part 1-2  Crm Cd  \
27751     19     Mission         1924         1     761   
134755     6   Hollywood          657         1     761   
363765     6   Hollywood          659         1     121   
490101     6   Hollywood          646         1     821   
537834    17  Devonshire         1797         1     122   
586003    14     Pacific         1452         2     910   
934749    19     Mission         1983         1     820   

                                                 Type Crm  ... Status  \
27751                                     BRANDISH WEAPON  ...     AA   
134755                                    BRANDISH WEAPON  ...     AO   
363765                                     RAPE, FORCIBLE  ...     IC   
490101  SODOMY/SEXUAL CONTACT B/W PENIS OF ONE PERS TO...  ...     IC   
537834                                    RAPE, ATTEMPTED  ...     AA   
586003                                         KIDNAPPING  ...     IC   
934749                                    ORAL COPULATION  ...     AO   

         Status Desc Crm Cd 1 Crm Cd 2  Crm Cd 3 Crm Cd 4  \
27751   Adult Arrest    761.0    930.0     997.0    998.0   
134755   Adult Other    761.0    920.0     930.0    998.0   
363765   Invest Cont    121.0    210.0     910.0    998.0   
490101   Invest Cont    230.0    821.0     910.0    998.0   
537834  Adult Arrest    122.0    230.0     910.0    998.0   
586003   Invest Cont    812.0    860.0     910.0    998.0   
934749   Adult Other    761.0    820.0     910.0    998.0   

                               LOCATION                     Cross Street  \
27751   ASTORIA                      ST  SAN FERNANDO                 RD   
134755                          WESTERN                          ROMAINE   
363765                        NORMANDIE                       DE LONGPRE   
490101                            SELMA                       LAS PALMAS   
537834  PARTHENIA                    ST                      HAYVENHURST   
586003                       WASHINGTON                         SPEEDWAY   
934749                           BURNET                        PARTHENIA   

            LAT       LON  
27751   34.2949 -118.4571  
134755  34.0885 -118.3092  
363765  34.0966 -118.3005  
490101  34.0997 -118.3363  
537834  34.2285 -118.4939  
586003  33.9792 -118.4666  
934749  34.2282 -118.4633  

[7 rows x 28 columns]